src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1318,29 +1469,29 @@ public static TSource Single<TSource>(this IEnumerable<TSource> source)
 
         public static TSource Single<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            long count = 0;
-            foreach (TSource element in source)
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (predicate(element))
+                while (e.MoveNext())
                 {
-                    result = element;
-                    checked { count++; }
+                    TSource result = e.Current;
+                    if (predicate(result))
+                    {
+                        while (e.MoveNext())
+                        {
+                            if (predicate(e.Current)) throw Error.MoreThanOneMatch();
+                        }
+                        return result;
+                    }
                 }
             }
-            switch (count)
-            {
-                case 0: throw Error.NoMatch();
-                case 1: return result;
-            }
-            throw Error.MoreThanOneMatch();
+            throw Error.NoMatch();
```

Comment:
I think I shall soon create a PR for a request that solely introduces short-circuits where appropriate, so that this issue can be discussed in isolation. Is that the best way to do that, in terms of procedures here?